0000-ilp-ping-protocol/0000-ilp-ping-protocol.md

```diff
@@ -0,0 +1,78 @@
+
+## Unidirectional Mode
+
+1. Initiator sends a ILP Prepare packet using the following details:
+   - **Destination**: Recipient's Interledger address.
+   - **Amount**: Any amount chosen by the Initiator.
+   - **Expiry**: Any appropriate expiry.
+   - **Condition**: The Known Condition, as defined in [Terminology](#terminology).
+   
+2. Upon receiving the Prepare packet, the Recipient MUST fulfill the payment using the following information:
+   - **Fulfillment**: The Known Fulfillment, as defined above.
+
+Note that a Recipient _MAY_ reject the payment if appropriate, for example due to an insufficient amount, invalid expiry, or for other factors.
+
+## Bidirectional Mode
+
+1. Initiator generates a new random 32-byte value to use as a Fulfilment (`F`) and generates the corresponding condition (`C`).
+2. Initiator sends an ILP Prepare packet to the Recipient with the following details:
+   - **Destination**: The Recipient's Interledger address.
+   - **Amount**: Any amount chosen by the Initiator.
+   - **Expiry**: An appropriate expiry that is long enough to allow for two round trips.
+   - **Condition**: The condition `C`.
+   - **Data**: A concatenated series of bytes with the following information:
+      - the bytes of the ASCII string `ECHOECHOECHOECHO`. In binary this value is `0x4543484F4543484F4543484F4543484F`.
+      - the byte `0x00`
+      - The ILP address of the Initiator (i.e., the return address) as an OER-encoded, variable length IA5 string.
+4. Upon receiving the Prepare packet, the Recipient identifies that it is a bidirectional Ping request by confirming the following:
+  1. The packet is addressed directly to itself (i.e. it has no tx or child suffix).
+  1. The packet has a payload conforming to the specification above (e.g., the byte following the `ECHOECHOECHOECHO` prefix in the payload is `0x00`).
+5. The Recipient DOES NOT immediately send an ILP Fulfill response. Instead the Recipient sends a new ILP `Prepare` packet addressed to the address parsed from the original original packet's payload. This new packet has following details:original 
+    - *Destination*: The Interledger address found in the original packet's data payload.
+    - *Amount*: Any amount chosen by the Initiator.
+    - *Expiry*: An expiry that is marginally smaller than the expiry on the original ILP Prepare.
+    - *Condition*: The same condition `C`.
+    - *Data*: A concatenated series of bytes with the following information:
+      - the bytes of the ASCII string `ECHOECHOECHOECHO` (i.e., `0x4543484F4543484F4543484F4543484F`).
+      - the byte `0x01`
+7. Upon receiving this packet, the Initiator identifies that it is a Pong by the fact that it is addressed to the address used in the original Ping request.
+8. The Initiator fulfills this second packet using the fulfilment `F`.
```

Comment:
Do they need to fulfill it? What does that tell them that they didn't already learn from getting the Prepare packet?

Reply:
Good question - I didn't invent this flow, so I'm speculating a bit here (there could be other rationale here that I'm overlooking), but if the Initiator doesn't fulfill the "pong", then the Receiver will likely _also_ reject the "ping", and so the packets won't actually have ever been paid for.

This isn't the end of the world because the amount of money earned from ping/pong requests should be negligible. However, as a matter of good Connector operation and ecosystem health, I feel pretty strongly that _every_ packet should allow a Connector to earn a fee (as much as possible - there are of course always Rejects). But as a matter of good protocol design, I prefer the mutual "fulfill" over a mutual "reject".

I suppose also if both sides reject, it's sort of unclear what _actually_ happened. From the perspective of the Initiator, that entity could _probably_ infer connectivity, but they wouldn't be able to see other things, like possibly a ping/pong rate....(?)